keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -53,51 +76,105 @@ class EnvironmentPromotionChecker(
                 }
               }
               else -> {
-                versions.firstOrNull { v ->
+                versions
+                  .firstOrNull { v ->
+                    pendingVersionsToCheck.remove(v)
+                    queuedForApproval.remove(v)
+                    if (vetoedVersions.contains(v)) {
+                      false
+                    } else {
+                      /**
+                       * Only check stateful evaluators if all stateless evaluators pass. We don't
+                       * want to request judgement or deploy a canary for artifacts that aren't
+                       * deployed to a required environment or outside of an allowed time.
+                       */
+                      val passesConstraints =
+                        checkConstraints(statelessEvaluators, artifact, deliveryConfig, v, environment) &&
+                          checkConstraints(statefulEvaluators, artifact, deliveryConfig, v, environment)
+
+                      versionIsPending = when (environment.constraints.anyStateful) {
+                        true -> repository
+                          .constraintStateFor(deliveryConfig.name, environment.name, v)
+                          .any { it.status == PENDING }
+                        else -> false
+                      }
+                      passesConstraints || versionIsPending
+                    }
+                  }
+              }
+            }
+
+            /**
+             * If there are pending constraints for prior versions, recheck in ascending version order
+             * so they can be timed out, failed, or approved. If a newer version was selected above,
+             * it is approved last in an attempt to retain ENVIRONMENT_ARTIFACT_VERSIONS.APPROVED_AT
+             * ordering.
+             */
+            var approvedPending = false
+            if (!hasPin) {
+              pendingVersionsToCheck
+                .sortedWith(artifact.versioningStrategy.comparator.reversed()) // oldest first
+                .forEach {
+                  if (checkConstraints(statefulEvaluators, artifact, deliveryConfig, it, environment)) {
```

Comment:
Why do we skip checking the stateless constraints here?

Reply:
Great catch, the comment bellow about why stateless constraints are re-checked for `queuedForApproval` versions does apply here as well. Will fix.